### PR TITLE
Replace Conv1d with Conv2d

### PR DESCRIPTION
**Pull request by taivu1998**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43221 Add benchmark for performance comparison
* **#42867 Replace Conv1d with Conv2d**

Differential Revision: [D23177916](https://our.internmc.facebook.com/intern/diff/D23177916)

### CODE REVIEW

File: test/test_xnnpack_integration.py [hunk 1]
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
...
+        deserialized_scripted_model = torch.jit.load(buffer)
+        for pattern, v in pattern_count_map.items():
+            if (v == 0):
+                FileCheck().check(pattern).run(deserialized_scripted_model.graph)
+            elif (v == -1):
+                FileCheck().check_not(pattern).run(deserialized_scripted_model.graph)
+            else:
+                FileCheck().check_count(pattern, v, exactly=True).run(deserialized_scripted_model.graph)
+        xnnpack_result = deserialized_scripted_model(input_data)
+        torch.testing.assert_allclose(ref_result, xnnpack_result, rtol=1e-2, atol=1e-3)
```
Comment (kimishpatel) [inline test/test_xnnpack_integration.py:735]: Also run the model through optimize_for_mobile pass and check if tests still pass. This is mainly to transform the `conv2d` to prepacked counterparts of it that use xnnpack.

File: test/test_xnnpack_integration.py [hunk 2]
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
...
+                super(Conv1D, self).__init__()
+                self.weight = torch.nn.Parameter(torch.Tensor(torch.rand(conv_weight_shape)), requires_grad=False)
+                self.bias = torch.nn.Parameter(torch.Tensor(torch.rand(conv_bias_shape)), requires_grad=False)
+                self.strides = strides
+                self.paddings = paddings
+                self.dilations = dilations
+                self.groups = groups
+
+            def forward(self, x):
+                return F.conv1d(x, self.weight, self.bias,
```
Comment (kimishpatel) [inline test/test_xnnpack_integration.py:772]: I would say let's do conv1d, relu, and then FC.

File: test/test_xnnpack_integration.py [hunk 3]
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
...
+
+    @given(batch_size=st.integers(0, 3),
+           input_channels_per_group=st.integers(1, 32),
+           height=st.integers(5, 64),
+           output_channels_per_group=st.integers(1, 32),
+           groups=st.integers(1, 16),
+           kernel=st.integers(1, 5),
+           strides=st.integers(1, 2),
+           paddings=st.integers(0, 2),
+           dilations=st.integers(1, 2))
```
Comment (kimishpatel) [inline test/test_xnnpack_integration.py:745]: I think test_xnnpack_integration.py needs to be fixed to not use hypothesis, because it generates only a handful of tests and no iterate through all comobinations.
If possible can we use https://docs.python.org/2/library/itertools.html#itertools.product, to get like all combinations? At least for some options like groups, kernel, strides, padding and dilations.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 4]
```diff
@@ -24,6 +24,30 @@ namespace jit {
...
 
+void replaceConv1dWithConv2d(std::shared_ptr<Graph>& graph) {
+  std::string conv_1d_pattern = R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+        %r = aten::conv1d(%input, %weight, %bias, %stride, %padding, %dilation, %groups)
+        return (%r) )";
+
+  std::string conv_2d_pattern = R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+        %padding_w : int = prim::ListUnpack(%padding)
```
Comment (kimishpatel) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:35]: Why only padding is adjusted like this? What about stride, and dilations?

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 5]
```diff
@@ -323,6 +363,16 @@ script::Module optimizeForMobile(
...
+void transformConv1dToConv2d(std::shared_ptr<Graph>& graph) {
+  TORCH_INTERNAL_ASSERT(
+      "XNNPACK is not enabled. Please build with USE_XNNPACK=1");
+}
+
+void transformConv1dToConv2d(script::Module& module) {
+  TORCH_INTERNAL_ASSERT(
+      "XNNPACK is not enabled. Please build with USE_XNNPACK=1");
+}
+
```
Comment (kimishpatel) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:375]: Actually, this is just rewriting graph into another graph that does not need xnnpack support to be executed. So we can remove these and move the other functions you added outside of ifdef.
Eventually we can think of moving them to a separate file once we have these working.

File: test/test_xnnpack_integration.py [hunk 6]
```diff
@@ -704,6 +704,78 @@ def forward(self, x):
         TestXNNPACKRewritePass.validate_transformed_module(DecomposedLinearMatmulAdd(), pattern_count_map, data_shape)
         TestXNNPACKRewritePass.validate_transformed_module(DecomposedLinearMatmul(), pattern_count_map, data_shape)
 
+@unittest.skipUnless(torch.backends.xnnpack.enabled,
+                     " XNNPACK must be enabled for these tests."
+                     " Please build with USE_XNNPACK=1.")
+class TestXNNPACKConv1dTransformPass(TestCase):
+    @staticmethod
+    def validate_transform_conv1d_to_conv2d(
```
Comment (kimishpatel) [inline test/test_xnnpack_integration.py:712]: I added this function in class because there was a reuse there. Since I dont see much reuse, you can inline this.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 7]
```diff
@@ -246,6 +270,22 @@ void runCanonicalOptimizations(script::Module& module) {
...
 
+void transformConv1dToConv2d(std::shared_ptr<Graph>& graph) {
+  // Replace _convolution with conv1d and conv2d
+  graph_rewrite_helper::replaceConvolutionWithAtenConv(graph);
+  replaceConv1dWithConv2d(graph);
+}
+
+void transformConv1dToConv2d(script::Module& module) {
+  for (auto& method : module.get_methods()) {
+    auto graph = method.graph();
```
Comment (kimishpatel) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:281]: `auto&` to avoid refcount bump.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Overall this looks good. Address some of the comments and think about adding benchmark. Good work!

File: test/test_xnnpack_integration.py (hunk 6 above)
Comment (taivu1998) [reply test/test_xnnpack_integration.py:712]: I thought we may need this function if we want to add more test cases later right?

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 4 above)
Comment (taivu1998) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:35]: I think we can handle stride and dilation similarly as well. However, it might not be needed since we only move in one dimension in conv1d, so the added dimension for stride and dilation may not matter too much. Padding is different, because we don't want to pad the added dimension.

File: test/test_xnnpack_integration.py (hunk 6 above)
Comment (kimishpatel) [reply test/test_xnnpack_integration.py:712]: Usually, we should introduce something only when we need it. But we can keep this for now. I am ok with it. You can add another test case that I described where we do conv1d+relu+FC etc.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:35]: I think we should be consistent and clear here. It is ok that conv2d might handle that gracefully, but it may change and not handle undefined value gracefully. We should avoid baking assumption about what some function might do, unless it is strongly coupled with our implementation.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 7 above)
Comment (taivu1998) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp]: I did try it, but it raised an error, so I just went for `auto`.

File: test/test_xnnpack_integration.py [hunk 8]
```diff
@@ -711,71 +713,163 @@ class TestXNNPACKConv1dTransformPass(TestCase):
...
+
+                def forward(self, x):
+                    return F.conv1d(x, self.weight, self.bias,
+                                    self.stride, self.padding, self.dilation, self.groups)
+
+            data_shape = (batch_size, input_channels, width)
+            pattern_count_transformed_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": 1}
+            pattern_count_optimized_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": -1}
```
Comment (kimishpatel) [inline test/test_xnnpack_integration.py]: Also add check for prepacked::conv2d_clamp_run or something like that. See other tests for the pattern.

File: test/test_xnnpack_integration.py [hunk 9]
```diff
@@ -711,71 +713,163 @@ class TestXNNPACKConv1dTransformPass(TestCase):
...
+
+                    self.fc_weight = torch.nn.Parameter(torch.Tensor(torch.rand(fc_weight_shape)), requires_grad=False)
+                    self.fc_bias = torch.nn.Parameter(torch.Tensor(torch.rand(fc_bias_shape)), requires_grad=False)
+
+                def forward(self, x):
+                    x = F.conv1d(x, self.conv_weight, self.conv_bias,
+                                 self.stride, self.padding, self.dilation, self.groups)
+                    x = F.relu(x)
+                    x = x.view(x.size(0), -1)
+                    x = F.linear(x, self.fc_weight, self.fc_bias)
```
Comment (kimishpatel) [inline test/test_xnnpack_integration.py:861]: Great. Thanks for adding this test.

File: test/test_xnnpack_integration.py [hunk 10]
```diff
@@ -711,71 +713,163 @@ class TestXNNPACKConv1dTransformPass(TestCase):
...
+                    return x
+
+            data_shape = (batch_size, input_channels, width)
+            pattern_count_transformed_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": 1}
+            pattern_count_optimized_map = {"Tensor = aten::conv1d": -1,
+                                            "Tensor = aten::conv2d": -1}
```
Comment (kimishpatel) [inline test/test_xnnpack_integration.py]: Similar to other comment. Add check for prepack stuff.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
This looks pretty good. Just one small change about adding additional check.

**Review by kimishpatel — APPROVED** [review]
